File: app/javascript/channels/chat_channel.js
/* global $*/
/* global location*/
import consumer from "./consumer"

if (!(location.pathname.includes('admin') || location.pathname == '/')) {
  // ChatChannelを購読し、WebSocket接続を確立
  const appChat = consumer.subscriptions.create("ChatChannel", {
    connected() {
      // サブスクリプションがサーバーで使用可能になったときに呼び出される
      //Called when the subscription is ready for use on the server
    },

    disconnected() {
      // サブスクリプションがサーバーによって終了されたときに呼び出される
      // Called when the subscription has been terminated by the server
    },

    received(data) {
       // WebSocketを介して受信したデータを処理
      let insertText = ""
      let my_image  = `<img src="${$("#user_image").attr('src')}">`;
      let partner_image = `<img src="${$("#partner_image").attr('src')}">`;
      let my_name = $("#current_user_name").val();
      let partner_name = $("#partner_name").val();
      let partner_id = $("#partner_id").val();
      // console.log(data,data.sentence)
      // 画面を開いているのがチャット送信者だった場合
      if (data["isCurrent_user"] == true) {

        insertText = `
            <div class="fukidasi">
              <div class="mycomment">
                <div class="text-right">
                  ${my_image}
                  <div class="name">${my_name}</div>
                </div>
                <p>${data.sentence}</p>
              </div>
            </div>
        `;
      }
      // 画面を開いているのがチャット受信者だった場合
      else if (partner_id == data.send_user) {
        insertText = `
            <div class="fukidasi">
              <div class="text-left">
                  ${partner_image}
              </div>
              <div class="chatting">
                <div class="name">
                  ${partner_name}
                </div>
                <div class="says">
                <p>${data.sentence}</p>
                </div>
              </div>
            </div>
        `;
      } else {
        return false;
      }
      // チャットメッセージを表示
      $("#chats").append(insertText);
      // チャットコンテナー内のスクロールを下に移動
      // Called when there's incoming data on the websocket for this channel
      setTimeout(function(){
        var objDiv = document.getElementById("chat_container");
        objDiv.scrollTop = objDiv.scrollHeight;
      },200)

    },

    speak(sentence) {
      // チャットメッセージを送信
      const current_user_id = $("#current_user_id").val();
      const partner_id = $("#partner_id").val();
      // console.log(sentence, "hoge2");
      return this.perform("speak", {
        sentence: sentence,
        current_user_id: current_user_id,
        partner_id: partner_id,
      });
    },
  });

  // メッセージ送信ボタンのクリックイベントハンドラ
  $(function () {
    $("#send").on("click", function (e) {
      const sentence = $("#sentence");
      // console.log(sentence, "hoge");
      appChat.speak(sentence.val());
      sentence.val(""); // フォームを空に
      e.preventDefault();
    });
  });
}

File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    # チャンネルに購読されたときの処理
    # "chat_channel" からのストリームを受信する
    stream_from "chat_channel"
    # 現在のユーザーのIDに基づいてストリームを受信する
    stream_for current_user.id
  end

  def unsubscribed
    # チャンネルが購読解除されたときの処理
    # 通常、クリーンアップ処理がここに含まれます
  end

  def speak(data)
    if data["sentence"]
      # チャットメッセージを作成してデータベースに保存
      Chat.create!(
        user_id: data["current_user_id"].to_i,
        partner_id: data["partner_id"].to_i,
        sentence: data["sentence"]
      )
      # 画面を開いているのがチャット送信者だった場合
      # チャットメッセージを送信者のチャンネルにブロードキャスト
      ChatChannel.broadcast_to data["current_user_id"].to_i,
        sentence: data["sentence"],
        partner_id: data["partner_id"],
        isCurrent_user: true

      # 画面を開いているのがチャット受信者だった場合
      # チャットメッセージを受信者のチャンネルにブロードキャスト
      ChatChannel.broadcast_to data["partner_id"].to_i,
        sentence: data["sentence"],
        partner_id: data["partner_id"],
        isCurrent_user: false,
        send_user: data["current_user_id"].to_i
    end
  end
end